File: apps/windows-electron/src/main/asr.js
'use strict';

// Offline ASR adapter for Windows using whisper.cpp. It shells out to a local
// whisper-cli executable and reads back the transcript. No cloud service is used.

const fs = require('fs');
const os = require('os');
const path = require('path');
const { execFile } = require('child_process');

const ROOT = path.resolve(__dirname, '..', '..', '..', '..');
const DEV_NATIVE_DIR = path.join(ROOT, 'native', 'asr');
const DEV_MODELS_DIR = path.join(ROOT, 'models');
const PACKAGED_NATIVE_DIR = path.join(process.resourcesPath || '', 'native', 'asr');
const PACKAGED_MODELS_DIR = path.join(process.resourcesPath || '', 'models');

const MODEL_FILES = {
  'tiny.en': 'ggml-tiny.en.bin',
  'base.en': 'ggml-base.en.bin',
  'small.en': 'ggml-small.en.bin',
};

function dedupe(paths) {
  return paths.filter((p, i) => p && paths.indexOf(p) === i);
}

function nativeDirs() {
  return dedupe([PACKAGED_NATIVE_DIR, DEV_NATIVE_DIR]);
}

function modelDirs() {
  return dedupe([PACKAGED_MODELS_DIR, DEV_MODELS_DIR]);
}

function findBinary() {
  const candidates = ['whisper-cli.exe', 'main.exe', 'whisper.exe'];
  for (const dir of nativeDirs()) {
    for (const c of candidates) {
      const p = path.join(dir, c);
      if (fs.existsSync(p)) return p;
    }
  }
  return null;
}

function modelPath(modelName) {
  const file = MODEL_FILES[modelName] || MODEL_FILES['base.en'];
  for (const dir of modelDirs()) {
    const p = path.join(dir, file);
    if (fs.existsSync(p)) return p;
  }
  return null;
}

function isReady(modelName = 'base.en') {
  return !!findBinary() && !!modelPath(modelName);
}

function setupMessage(modelName = 'base.en') {
  const file = MODEL_FILES[modelName] || MODEL_FILES['base.en'];
  return (
    'Offline ASR is not set up yet.\n' +
    `1. Put a whisper.cpp build (whisper-cli.exe) in: ${nativeDirs().join(' or ')}\n` +
    `2. Put the model "${file}" in: ${modelDirs().join(' or ')}\n` +
    'No internet is used - everything runs locally.'
  );
}

function toBuffer(pcm16) {
  if (Buffer.isBuffer(pcm16)) return pcm16;
  if (ArrayBuffer.isView(pcm16)) {
    return Buffer.from(pcm16.buffer, pcm16.byteOffset, pcm16.byteLength);
  }
  return Buffer.from(pcm16);
}

function writeWav(pcm16, sampleRate = 16000) {
  const data = toBuffer(pcm16);
  const header = Buffer.alloc(44);
  header.write('RIFF', 0);
  header.writeUInt32LE(36 + data.length, 4);
  header.write('WAVE', 8);
  header.write('fmt ', 12);
  header.writeUInt32LE(16, 16);
  header.writeUInt16LE(1, 20);
  header.writeUInt16LE(1, 22);
  header.writeUInt32LE(sampleRate, 24);
  header.writeUInt32LE(sampleRate * 2, 28);
  header.writeUInt16LE(2, 32);
  header.writeUInt16LE(16, 34);
  header.write('data', 36);
  header.writeUInt32LE(data.length, 40);
  const file = path.join(os.tmpdir(), `coldvoice-${Date.now()}.wav`);
  fs.writeFileSync(file, Buffer.concat([header, data]));
  return file;
}

function runWhisper(wavFile, modelName) {
  return new Promise((resolve, reject) => {
    const bin = findBinary();
    const model = modelPath(modelName);
    if (!bin || !model) return reject(new Error(setupMessage(modelName)));
    const args = ['-m', model, '-f', wavFile, '-nt', '-otxt', '-of', wavFile];
    execFile(bin, args, { timeout: 120000 }, (err) => {
      if (err) return reject(err);
      const txtFile = `${wavFile}.txt`;
      try {
        const text = stripNonSpeech(fs.readFileSync(txtFile, 'utf8'));
        cleanup([wavFile, txtFile]);
        resolve(text);
      } catch (e) {
        reject(e);
      }
    });
  });
}

function stripNonSpeech(text) {
  return String(text)
    .split(/\r?\n/)
    .map((l) => l.replace(/\[[^\]]*\]/g, '').replace(/\((?:blank_audio|silence|inaudible|music|pause|noise)\)/gi, '').trim())
    .filter(Boolean)
    .join(' ')
    .trim();
}

function cleanup(files) {
  for (const f of files) {
    try {
      fs.unlinkSync(f);
    } catch {
      /* ignore */
    }
  }
}

async function transcribe(pcm16, modelName = 'base.en', sampleRate = 16000) {
  const wav = writeWav(pcm16, sampleRate);
  try {
    return await runWhisper(wav, modelName);
  } finally {
    cleanup([wav]);
  }
}

module.exports = {
  transcribe,
  isReady,
  setupMessage,
  get MODELS_DIR() { return modelDirs()[0]; },
  get NATIVE_DIR() { return nativeDirs()[0]; },
  modelDirs,
  nativeDirs,
};

File: apps/linux-tauri/src-tauri/src/asr.rs
// Offline ASR adapter for Linux using whisper.cpp. It shells out to a local
// whisper-cli executable and reads back the transcript. No cloud service is
// used. The binary and the model ship inside the .deb / AppImage, and a user
// can drop their own build or a bigger model into ~/.local/share/coldvoice.

use once_cell::sync::OnceCell;
use std::path::{Path, PathBuf};
use std::process::Command;

use crate::util;

static RESOURCE_DIR: OnceCell<PathBuf> = OnceCell::new();

pub fn init(resource_dir: PathBuf) {
    let _ = RESOURCE_DIR.set(resource_dir);
    // Bundlers do not always preserve the executable bit on resources.
    if let Some(bin) = find_binary() {
        make_executable(&bin);
    }
}

fn make_executable(path: &Path) {
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        if let Ok(meta) = std::fs::metadata(path) {
            let mode = meta.permissions().mode();
            if mode & 0o111 == 0 {
                let _ = std::fs::set_permissions(path, std::fs::Permissions::from_mode(mode | 0o755));
            }
        }
    }
}

pub fn model_file(model_name: &str) -> &'static str {
    match model_name {
        "tiny.en" => "ggml-tiny.en.bin",
        "small.en" => "ggml-small.en.bin",
        _ => "ggml-base.en.bin",
    }
}

pub fn native_dirs() -> Vec<PathBuf> {
    let mut dirs = Vec::new();
    if let Some(res) = RESOURCE_DIR.get() {
        dirs.push(res.join("native").join("asr"));
    }
    dirs.push(util::data_dir().join("native").join("asr"));
    dirs
}

pub fn model_dirs() -> Vec<PathBuf> {
    let mut dirs = Vec::new();
    if let Some(res) = RESOURCE_DIR.get() {
        dirs.push(res.join("models"));
    }
    dirs.push(util::data_dir().join("models"));
    dirs
}

pub fn find_binary() -> Option<PathBuf> {
    for dir in native_dirs() {
        for name in ["whisper-cli", "main", "whisper"] {
            let p = dir.join(name);
            if p.is_file() {
                return Some(p);
            }
        }
    }
    // A system-wide whisper.cpp build is a perfectly good fallback.
    if let Ok(path) = std::env::var("PATH") {
        for dir in path.split(':') {
            let p = Path::new(dir).join("whisper-cli");
            if p.is_file() {
                return Some(p);
            }
        }
    }
    None
}

pub fn model_path(model_name: &str) -> Option<PathBuf> {
    let file = model_file(model_name);
    for dir in model_dirs() {
        let p = dir.join(file);
        if p.is_file() {
            return Some(p);
        }
    }
    None
}

pub fn is_ready(model_name: &str) -> bool {
    find_binary().is_some() && model_path(model_name).is_some()
}

pub fn setup_message(model_name: &str) -> String {
    let file = model_file(model_name);
    format!(
        "Offline ASR is not set up yet.\n1. Put a whisper.cpp build (whisper-cli) in: {}\n2. Put the model \"{}\" in: {}\nNo internet is used - everything runs locally.",
        native_dirs().iter().map(|p| p.display().to_string()).collect::<Vec<_>>().join(" or "),
        file,
        model_dirs().iter().map(|p| p.display().to_string()).collect::<Vec<_>>().join(" or "),
    )
}

fn strip_non_speech(text: &str) -> String {
    // Strip whisper's non-speech annotations: [BLANK_AUDIO], (music), (wind
    // blowing), etc. These are always fully bracketed or parenthesised.
    let mut lines: Vec<String> = Vec::new();
    for line in text.lines() {
        let cleaned = strip_bracketed(line);
        let cleaned = cleaned.trim();
        if !cleaned.is_empty() {
            lines.push(cleaned.to_string());
        }
    }
    lines.join(" ").trim().to_string()
}

fn strip_bracketed(line: &str) -> String {
    let mut out = String::with_capacity(line.len());
    let mut square = 0i32;
    let mut round = 0i32;
    for c in line.chars() {
        match c {
            '[' => square += 1,
            ']' => square = (square - 1).max(0),
            '(' => round += 1,
            ')' => round = (round - 1).max(0),
            _ => {
                if square == 0 && round == 0 {
                    out.push(c);
                }
            }
        }
    }
    out
}

fn cpu_threads() -> usize {
    std::thread::available_parallelism().map(|n| n.get()).unwrap_or(1).max(1)
}

// Transcribe 16 kHz mono PCM16 with the local model. Blocking: callers run this
// on a worker thread, never on the UI thread.
pub fn transcribe(pcm: &[i16], model_name: &str, sample_rate: u32) -> Result<String, String> {
    let bin = find_binary().ok_or_else(|| setup_message(model_name))?;
    let model = model_path(model_name).ok_or_else(|| setup_message(model_name))?;

    let wav_path = std::env::temp_dir().join(format!("coldvoice-{}.wav", util::now_ms()));
    std::fs::write(&wav_path, util::wav_buffer(pcm, sample_rate)).map_err(|e| e.to_string())?;
    let txt_path = PathBuf::from(format!("{}.txt", wav_path.display()));

    // Speed flags: all CPU cores, greedy decode (beam size 1 / best-of 1), no
    // temperature fallback. Same set the Windows build uses.
    let status = Command::new(&bin)
        .arg("-m").arg(&model)
        .arg("-f").arg(&wav_path)
        .arg("-t").arg(cpu_threads().to_string())
        .arg("-bs").arg("1")
        .arg("-bo").arg("1")
        .arg("-nf")
        .arg("-nt")
        .arg("-otxt")
        .arg("-of").arg(&wav_path)
        .output();

    let result = match status {
        Ok(out) if out.status.success() => std::fs::read_to_string(&txt_path)
            .map(|t| strip_non_speech(&t))
            .map_err(|e| e.to_string()),
        Ok(out) => Err(format!(
            "whisper-cli exited with {}: {}",
            out.status,
            String::from_utf8_lossy(&out.stderr).chars().take(300).collect::<String>()
        )),
        Err(e) => Err(e.to_string()),
    };

    let _ = std::fs::remove_file(&wav_path);
    let _ = std::fs::remove_file(&txt_path);
    result
}
